web/packages/shared/hooks/useInfiniteScroll/useKeyBasedPagination.ts

```diff
@@ -72,9 +70,11 @@ export function useKeyBasedPagination<T>({
   // https://react.dev/learn/you-might-not-need-an-effect#adjusting-some-state-when-a-prop-changes,
   // we can ensure that the state is reset before anything renders, and thereby
   // ensure that the new `fetch` function is used.
-  const [prevFilter, setPrevFilter] = useState(filter);
-  if (prevFilter !== filter) {
-    setPrevFilter(filter);
+  const [prevFetchFunc, setPrevFetchFunc] = useState(() => fetchFunc);
+  if (prevFetchFunc !== fetchFunc) {
+    // we have to use the callback form because we want to store a function
+    // in the state, not the returned value
+    setPrevFetchFunc(() => fetchFunc);
     clear();
```

Comment:
The overall behavior related to clearing the state seems to indicate that maybe `useKeyBasedPagination` takes care of too much stuff. To my understanding, the general behavior we want is that whenever we change filters, we want to clear the stored resources and start fresh.

The current solution binds this behavior to the change in identity of `fetchFunc` argument, which in case of `teleport/src/UnifiedResources/UnifiedResources.tsx` is bound to changing the filters. This should work for now, but in the future `useCallback` in `teleport/src/UnifiedResources/UnifiedResources.tsx` might include some extra deps which shouldn't trigger a full clear if those deps change.

`useKeyBasedPagination.fetch` also enables clearing the state through the extra `fromStart` argument. It seems to me that instead, `useKeyBasedPagination` should return the `clear` function and let the consumer control when the state resets. The consumer could then clear the state whenever the filters change etc.

This is not a request to change this piece of code, more like just thinking out loud in case someone else had this thought when working on this part of the codebase.

Reply:
Returning a `clear` function that would be called explicitly sounds like a great idea to me, and I'd love to hear others' thoughts.

Reply:
What @ravicious described sounds exactly what my thought train looked like when I was designing this part. The reason why I decided to do it like this is that I did it the TDD way and when I started to think "OK, how should this code behave if I change filters, but don't ask it to fetch from the start", it became clear that the "clear and start over" function is simply redundant.

Reply:
(That being said, of course, the "from start" flag is necessary to solve your use case.)

Reply:
That being said, of course, now we have another use case, and such a feature is needed. I also think that a `clear` function is a good idea. It gives you more flexibility over when and where you decide to start over.

Reply:
>I started to think "OK, how should this code behave if I change filters, but don't ask it to fetch from the start"

I see, however, the filters were not really used in the `useKeyBasedPagination` hook, so probably we shouldn't even pass them there. They were taken as an argument and passed on to the fetch function, but both the hook and the fetch function didn't need them. Because of that, I think that the idea of returning the `clear` function to the consumer which also keeps the params is more "clean".

-------------------

I thought that I would be able to get rid of this state adjusting mechanism:
```
  const [prevParams, setPrevParams] = useState(params);
  if (prevParams !== params) {
    setPrevParams(params);
    clear();
  }
```
but I had to re-use it in `teleport/UnifiedResources`. In the Web UI we get params from `useUrlFiltering` hook, so we don't have a full control over when they are changed. To remedy this, we could add a `onParamsChange` callback to it, that would be called when the internal `setParams` is called. 
OTOH if we would like to refactor the hook in the future to use [`useSearchParams`](https://reactrouter.com/en/main/hooks/use-search-params) (to make the URL the only source of truth, instead of keeping the params in `useState`) it would be more difficult to call that callback when the URL query params change, since `useSearchParams` provides just the updated value, without any listener. 

I think I will keep it as is, in Connect we will be able to call `clear` when `setParams` is called.